rclcpp/test/rclcpp/executors/test_priority_events_executor.cpp

```diff
@@ -0,0 +1,163 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <gtest/gtest.h>
+
+#include <chrono>
+#include <memory>
+#include <string>
+
+#include "rclcpp/experimental/executors/events_executor/events_executor.hpp"
+#include "rclcpp/experimental/executors/events_executor/priority_events_queue.hpp"
+// #include "rclcpp/experimental/executors/events_executor/simple_events_queue.hpp"
+
+#include "test_msgs/srv/empty.hpp"
+#include "test_msgs/msg/empty.hpp"
+
+using namespace std::chrono_literals;
+
+using rclcpp::experimental::executors::EventsExecutor;
+using rclcpp::experimental::executors::PriorityEventsQueue;
+// using rclcpp::experimental::executors::SimpleEventsQueue;
+
+class TestPriorityEventsExecutor : public ::testing::Test
+{
+public:
+  void SetUp()
+  {
+    rclcpp::init(0, nullptr);
+  }
+
+  void TearDown()
+  {
+    rclcpp::shutdown();
+  }
+};
+
+TEST_F(TestPriorityEventsExecutor, priority_subs)
+{
+  // rmw_connextdds doesn't support events-executor
+  if (std::string(rmw_get_implementation_identifier()).find("rmw_connextdds") == 0) {
+    GTEST_SKIP();
+  }
+
+  // Create timer publishing to 3 subscriptions
+  auto node = std::make_shared<rclcpp::Node>("node");
+
+  // Create QoS for subscriptions
+  rclcpp::QoS qos(10);
+
+  int subscriptions_executed = 0;
+
+  qos.deadline(rclcpp::Duration(3, 0));
+  bool msg_received_low = false;
+  auto sub_low = node->create_subscription<test_msgs::msg::Empty>(
+    "topic", qos,
+    [&msg_received_low, &subscriptions_executed](test_msgs::msg::Empty::ConstSharedPtr msg)
+    {
+      (void)msg;
+      msg_received_low = true;
+      EXPECT_EQ(subscriptions_executed, 2);
+      subscriptions_executed++;
+    });
+
+  qos.deadline(rclcpp::Duration(2, 0));
+  bool msg_received_medium = false;
+  auto sub_medium = node->create_subscription<test_msgs::msg::Empty>(
+    "topic", qos,
+    [&msg_received_medium, &subscriptions_executed](test_msgs::msg::Empty::ConstSharedPtr msg)
+    {
+      (void)msg;
+      msg_received_medium = true;
+      EXPECT_EQ(subscriptions_executed, 1);
+      subscriptions_executed++;
+    });
+
+  qos.deadline(rclcpp::Duration(1, 0));
+  bool msg_received_high = false;
+  auto sub_high = node->create_subscription<test_msgs::msg::Empty>(
+    "topic", qos,
+    [&msg_received_high, &subscriptions_executed](test_msgs::msg::Empty::ConstSharedPtr msg)
+    {
+      (void)msg;
+      msg_received_high = true;
+      EXPECT_EQ(subscriptions_executed, 0);
+      subscriptions_executed++;
+    });
+
+  auto publisher = node->create_publisher<test_msgs::msg::Empty>("topic", qos);
+
+  // Create executor
+  auto extract_priority = [](const rclcpp::experimental::executors::ExecutorEvent & event) {
```

Comment:
It looks like for the sake of this unit-test, this function can be greatly simplified to
```
if (event.type != rclcpp::experimental::executors::ExecutorEventType::SUBSCRIPTION_EVENT) {
    return 0L;
}
auto subscription = static_cast<const rcl_subscription_t *>(event.entity_key);
return rcl_subscription_get_options(subscription)->qos.deadline.sec;
```